// angular-bootstrap-md/esm2015/lib/free/dropdown/dropdown-container.component.js
import { __decorate, __metadata } from "tslib";
import { ChangeDetectionStrategy, Component, HostBinding } from '@angular/core';
import { BsDropdownState } from './dropdown.state';
let BsDropdownContainerComponent = class BsDropdownContainerComponent {
    constructor(_state) {
        this._state = _state;
        this.isOpen = false;
        this.display = 'block';
        this.position = 'absolute';
        this._subscription = _state.isOpenChange.subscribe((value) => {
            this.isOpen = value;
        });
    }
    get direction() {
        return this._state.direction;
    }
    ngOnDestroy() {
        this._subscription.unsubscribe();
    }
};
BsDropdownContainerComponent.ctorParameters = () => [
    { type: BsDropdownState }
];
__decorate([
    HostBinding('style.display'),
    __metadata("design:type", Object)
], BsDropdownContainerComponent.prototype, "display", void 0);
__decorate([
    HostBinding('style.position'),
    __metadata("design:type", Object)
], BsDropdownContainerComponent.prototype, "position", void 0);
BsDropdownContainerComponent = __decorate([
    Component({
        selector: 'mdb-dropdown-container',
        changeDetection: ChangeDetectionStrategy.OnPush,
        template: `
    <div
      [class.dropup]="direction === 'up'"
      [class.dropdown]="direction === 'down'"
      [class.show]="isOpen"
      [class.open]="isOpen"
    >
      <ng-content></ng-content>
    </div>
  `
    }),
    __metadata("design:paramtypes", [BsDropdownState])
], BsDropdownContainerComponent);
export { BsDropdownContainerComponent };
