// angular-bootstrap-md/esm2015/lib/free/dropdown/dropdown.module.js
var DropdownModule_1;
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { ComponentLoaderFactory } from '../utils/component-loader/component-loader.factory';
import { PositioningService } from '../utils/positioning/positioning.service';
import { BsDropdownContainerComponent } from './dropdown-container.component';
import { BsDropdownMenuDirective } from './dropdown-menu.directive';
import { BsDropdownToggleDirective } from './dropdown-toggle.directive';
import { BsDropdownConfig } from './dropdown.config';
import { BsDropdownDirective } from './dropdown.directive';
import { BsDropdownState } from './dropdown.state';
let DropdownModule = DropdownModule_1 = class DropdownModule {
    static forRoot(config) {
        return {
            ngModule: DropdownModule_1,
            providers: [
                ComponentLoaderFactory,
                PositioningService,
                BsDropdownState,
                { provide: BsDropdownConfig, useValue: config ? config : { autoClose: true } },
            ],
        };
    }
};
DropdownModule = DropdownModule_1 = __decorate([
    NgModule({
        declarations: [
            BsDropdownMenuDirective,
            BsDropdownToggleDirective,
            BsDropdownContainerComponent,
            BsDropdownDirective,
        ],
        exports: [BsDropdownMenuDirective, BsDropdownToggleDirective, BsDropdownDirective],
        entryComponents: [BsDropdownContainerComponent],
    })
], DropdownModule);
export { DropdownModule };
