// angular-bootstrap-md/esm2015/lib/free/modals/modal.module.js
var ModalModule_1;
import { __decorate } from "tslib";
import { NgModule, NO_ERRORS_SCHEMA } from '@angular/core';
import { ModalBackdropComponent } from './modalBackdrop.component';
import { ModalDirective } from './modal.directive';
import { PositioningService } from '../utils/positioning/positioning.service';
import { ComponentLoaderFactory } from '../utils/component-loader/component-loader.factory';
import { ModalContainerComponent } from './modalContainer.component';
import { MDBModalService } from './modal.service';
import { A11yModule } from '@angular/cdk/a11y';
let ModalModule = ModalModule_1 = class ModalModule {
    static forRoot() {
        return {
            ngModule: ModalModule_1,
            providers: [MDBModalService, ComponentLoaderFactory, PositioningService],
        };
    }
};
ModalModule = ModalModule_1 = __decorate([
    NgModule({
        declarations: [ModalBackdropComponent, ModalDirective, ModalContainerComponent],
        imports: [A11yModule],
        exports: [ModalBackdropComponent, ModalDirective],
        entryComponents: [ModalBackdropComponent, ModalContainerComponent],
        schemas: [NO_ERRORS_SCHEMA],
    })
], ModalModule);
export { ModalModule };
