// angular-bootstrap-md/esm2015/lib/free/navbars/navlinks.component.js
import { __decorate, __metadata } from "tslib";
import { NavbarService } from './navbar.service';
import { AfterContentInit, Component, ContentChildren, ElementRef, QueryList, EventEmitter, Output, Renderer2, } from '@angular/core';
import { RouterLinkWithHref } from '@angular/router';
let NavlinksComponent = class NavlinksComponent {
    constructor(_navbarService, renderer) {
        this._navbarService = _navbarService;
        this.renderer = renderer;
        this.linkClick = new EventEmitter();
    }
    ngAfterContentInit() {
        setTimeout(() => {
            this.links.forEach((link) => {
                this.renderer.listen(link.nativeElement, 'click', () => {
                    this._navbarService.setNavbarLinkClicks();
                });
            });
        }, 0);
    }
};
NavlinksComponent.ctorParameters = () => [
    { type: NavbarService },
    { type: Renderer2 }
];
__decorate([
    ContentChildren(RouterLinkWithHref, { read: ElementRef, descendants: true }),
    __metadata("design:type", QueryList)
], NavlinksComponent.prototype, "links", void 0);
__decorate([
    Output(),
    __metadata("design:type", Object)
], NavlinksComponent.prototype, "linkClick", void 0);
NavlinksComponent = __decorate([
    Component({
        // tslint:disable-next-line:component-selector
        selector: 'navlinks',
        template: `
    <ng-content></ng-content>
  `
    }),
    __metadata("design:paramtypes", [NavbarService, Renderer2])
], NavlinksComponent);
export { NavlinksComponent };
