// angular-bootstrap-md/esm2015/lib/free/popover/popover.module.js
var PopoverModule_1;
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ComponentLoaderFactory } from '../utils/component-loader/component-loader.factory';
import { PositioningService } from '../utils/positioning/positioning.service';
import { PopoverConfig } from './popover.config';
import { PopoverDirective } from './popover.directive';
import { PopoverContainerComponent } from './popover-container.component';
let PopoverModule = PopoverModule_1 = class PopoverModule {
    static forRoot() {
        return {
            ngModule: PopoverModule_1,
            providers: [PopoverConfig, ComponentLoaderFactory, PositioningService],
        };
    }
};
PopoverModule = PopoverModule_1 = __decorate([
    NgModule({
        imports: [CommonModule],
        declarations: [PopoverDirective, PopoverContainerComponent],
        exports: [PopoverDirective],
        entryComponents: [PopoverContainerComponent],
    })
], PopoverModule);
export { PopoverModule };
